File: BLIND-75/Array/maximum-subarray.ts
function maxSubArray(nums: number[]): number {
  let currSum = 0;
  let bestSum = -Infinity;
  
  for (let num of nums) {
      currSum = currSum+num > num ? currSum+num : num;
      bestSum = bestSum > currSum ? bestSum : currSum;
  }
  
  return bestSum;
}

File: restart_prep/maximum_subarray.cpp
#include <algorithm>
#include <cmath>
#include <vector>

/*
 *Given an integer array nums, find the 
subarray
 with the largest sum, and return its sum.
 * */

/*
 *
 * Input: nums = [-2,1,-3,4,-1,2,1,-5,4]
Output: 6
Explanation: The subarray [4,-1,2,1] has the largest sum 6.
*/
class Solution {
public:
    int maxSubArray(std::vector<int>& nums) {
		int maxSumSoFar = -100000000;
		int currSum = 0;
		for (int i = 0; i < nums.size(); i++) {
			currSum += nums[i];
			maxSumSoFar = std::max(std::max(currSum,nums[i]), maxSumSoFar);
			if (currSum < nums[i]) {
				currSum = nums[i];
			}
		}

		return maxSumSoFar;
    }
};
